src/main/java/net/minestom/server/inventory/AbstractInventory.java

```diff
@@ -0,0 +1,362 @@
+    }
+
+    /**
+     * Takes {@link ItemStack}s from the inventory and sends relevant updates to the viewer(s).
+     * <p>
+     * Even items cannot be fully taken, the amount of {@code itemStack}s will be updated.
+     *
+     * @param itemStacks items to take
+     * @return list of itemstacks that could not be successfully fully taken, empty list otherwise
+     */
+    public @NotNull List<ItemStack> takeItemStacks(@NotNull List<ItemStack> itemStacks) {
+        List<ItemStack> result = new ArrayList<>();
+        itemStacks.forEach(itemStack -> {
+            if (!takeItemStack(itemStack)) {
+                result.add(itemStack);
+            }
+        });
+        return result;
+    }
+
+    /**
+     * Checks whether {@link ItemStack} can be fully taken from the inventory.
+     *
+     * @param itemStack the item to be checked
+     * @return true if the item can be fully taken from the inventory, false otherwise
+     */
+    public boolean canTakeItemStack(@NotNull ItemStack itemStack) {
+        final StackingRule stackingRule = itemStack.getStackingRule();
+        int amountLeft = stackingRule.getAmount(itemStack);
+        for (int i = 0; i < getInnerSize(); i++) {
+            ItemStack inventoryItem = getItemStack(i);
+            if (inventoryItem.isAir()) {
+                continue;
+            }
+            if (stackingRule.canBeStacked(itemStack, inventoryItem)) {
+                final int itemAmount = stackingRule.getAmount(inventoryItem);
+                if (amountLeft <= itemAmount) {
+                    return true;
+                }
+                amountLeft -= itemAmount;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether {@link ItemStack}s can be fully taken from the inventory.
+     *
+     * @param itemStacks items to be checked
+     * @return true if all the items can be fully taken from the inventory, false otherwise
+     */
+    public boolean canTakeItemStacks(@NotNull List<ItemStack> itemStacks) {
+        return itemStacks.stream().allMatch(this::canTakeItemStack);
+    }
+
+    public synchronized void replaceItemStack(int slot, @NotNull UnaryOperator<@NotNull ItemStack> operator) {
+        var currentItem = getItemStack(slot);
+        setItemStack(slot, operator.apply(currentItem));
+    }
+
+    /**
+     * Clears the inventory and send relevant update to the viewer(s).
+     */
+    public synchronized void clear() {
+        // Clear the item array
+        Arrays.fill(itemStacks, ItemStack.AIR);
+        // Send the cleared inventory to viewers
+        update();
+    }
+
+    public abstract void update();
+
+    /**
+     * Gets the {@link ItemStack} at the specified slot.
+     *
+     * @param slot the slot to check
+     * @return the item in the slot {@code slot}
+     */
+    public @NotNull ItemStack getItemStack(int slot) {
+        return itemStacks[slot];
+    }
+
+    /**
+     * Gets all the {@link ItemStack} in the inventory.
```

Comment:
Might be worth noting that this specifically returns a clone of the items in the inventory. Perhaps with an `@see` or a note mentioning that getting each individual item stack would be considerably faster.

We could even have a `.forEachItem(ObjIntConsumer<ItemStack>)` method that allows people to iterate through an inventory without having to clone the array? In this case, the int would be the "slot" number.